src/tbb/threading_control.cpp

```diff
@@ -0,0 +1,397 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include "threading_control.h"
+#include "permit_manager.h"
+#include "market.h"
+#include "thread_dispatcher.h"
+#include "governor.h"
+#include "thread_dispatcher_client.h"
+
+namespace tbb {
+namespace detail {
+namespace r1 {
+
+// ---------------------------------------- threading_control_impl --------------------------------------------------------------
+
+std::size_t global_control_active_value_unsafe(d1::global_control::parameter);
+
+std::pair<unsigned, unsigned> threading_control_impl::calculate_workers_limits() {
+    // Expecting that 4P is suitable for most applications.
+    // Limit to 2P for large thread number.
+    // TODO: ask RML for max concurrency and possibly correct hard_limit
+    unsigned factor = governor::default_num_threads() <= 128 ? 4 : 2;
+
+    // The requested number of threads is intentionally not considered in
+    // computation of the hard limit, in order to separate responsibilities
+    // and avoid complicated interactions between global_control and task_scheduler_init.
+    // The threading control guarantees that at least 256 threads might be created.
+    unsigned workers_app_limit = global_control_active_value_unsafe(global_control::max_allowed_parallelism);
+    unsigned workers_hard_limit = max(max(factor * governor::default_num_threads(), 256u), workers_app_limit);
+    unsigned workers_soft_limit = calc_workers_soft_limit(governor::default_num_threads(), workers_hard_limit);
+    
+    return std::make_pair(workers_soft_limit, workers_hard_limit);
+}
+
+unsigned threading_control_impl::calc_workers_soft_limit(unsigned workers_soft_limit, unsigned workers_hard_limit) {
+    unsigned soft_limit = global_control_active_value_unsafe(global_control::max_allowed_parallelism);
+    if (soft_limit) {
+        workers_soft_limit = soft_limit - 1;
+    } else {
+        // if user set no limits (yet), use threading control's parameter
+        workers_soft_limit = max(governor::default_num_threads() - 1, workers_soft_limit);
+    }
+
+    if (workers_soft_limit >= workers_hard_limit) {
+        workers_soft_limit = workers_hard_limit - 1;
+    }
+
+    return workers_soft_limit;
+}
+
+d1::cache_aligned_unique_ptr<permit_manager> threading_control_impl::make_permit_manager(unsigned workers_soft_limit) {
+    return d1::make_cache_aligned_unique<market>(workers_soft_limit);
+}
+
+d1::cache_aligned_unique_ptr<thread_dispatcher> threading_control_impl::make_thread_dispatcher(threading_control& tc,
+                                                                                               unsigned workers_soft_limit,
+                                                                                               unsigned workers_hard_limit)
```

Comment:
Why did you split on 114th column then? Also, the alignment of the arguments has been broken.

The problem here is that everyone has their own screens. Desktop monitors usually have wider screens than laptop screens. However, even on wider screens people prefer having two or more files open side-by-side, which diminishes the width advantage of the screens. In addition, for larger monitors people increase font size to reduce eye focus exertion when reading the code. Many people also use vertical screens.

One of the extremes is 120 line length. However, even that many characters make the lines wrap when doing side-by-side review on the GitHub using 2.5k resolution, which is the most widely used display nowadays. 
Another extreme is the margin of 72 characters. This is used to be the maximum line length because old-school terminal windows have 72 characters, and punch cards had 80 characters. So, if we frequently read the code on such screens, it makes sense to follow that length of lines. Fortunately, we are not. Unfortunately, we are not aware about how others people read it. After all, we are writing not only for yourself.
I found many proofs on the Internet that people still prefer using 80 characters limit. Here are a couple of them: [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Line_Length) keeps that limit describing pros and cons of that decision, Linux kernel [used to enforce that limit](https://www.phoronix.com/news/Linux-Kernel-Deprecates-80-Col), but now only recommends sticking to it.

I believe there should be some reasonable tradeoff. Let's try to find it. I think the most important aspects of the line length are:
- Be able to read the code continuously without side scrolling of the screen. The horizontal/side scrolling distracts reader's attention. Especially, if the amount of side scrolling is huge. Consider doing horizontal scrolling back and forth for every line of code.
- Limiting complexity of a single line
- Limiting indentation, which in turn limiting the cyclomatic complexity by prohibiting too much of a conditional blocks, loops, and other type of  nesting.

For the 2.5K desktop monitor and FHD laptop the golden mean is ~100 characters.

Reply:
Not only this patch is affected but a bunch of different files. Do you agree to move such improvement to separate activity?